src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,403 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <xiaoyuan.song@intel.com>
+
+#include <ipc4/mixin_mixout.h>
+#include <sof/common.h>
+
+#ifdef MIXIN_MIXOUT_GENERIC
+
+#if CONFIG_FORMAT_S16LE
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function
+ * to also mix an entire stream. In this case the function is called with fake stream
+ * parameters: multichannel stream is treated as single channel and so the entire stream
+ * contents is mixed.
+ */
+static void normal_mix_channel_s16(struct audio_stream __sparse_cache *sink, uint32_t start_frame,
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	int16_t *dst = (int16_t *)sink->w_ptr + start_frame;
+	int16_t *src = (int16_t *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = sat_int16((int32_t)*dst + (int32_t)*src++);
+			dst++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++)
+			*dst++ = *src++;
+	}
+}
+
+static void remap_mix_channel_s16(struct audio_stream __sparse_cache *sink,
+				  uint8_t sink_channel_index,  uint8_t sink_channel_count,
+				  uint32_t start_frame, uint32_t mixed_frames,
+				  const struct audio_stream __sparse_cache *source,
+				  uint8_t source_channel_index, uint8_t source_channel_count,
```

Comment:
These uint8_t types are used in the hot loop code. Have you tried if change to int improves code speed? Xtensa is slow with 8 bit types and also I don't see specific need for unsigned.

Reply:
With lots of parameters to function a struct reference may be faster?

Reply:
> and also I don't see specific need for unsigned.

One of the big differences between signed and unsigned is signed has no well-defined overflow and underflow. I don't know whether overflow matters here.

But the worst things that can happen (including security issues) is when mixing signed and unsigned together.

Reply:
> With lots of parameters to function a struct reference may be faster?

If use a struct, we need to assign value to it before we call the function. And I found that not in all cases, the fewer parameters, the better, like https://github.com/thesofproject/sof/pull/6556. For example, test by xtensa simulator, if I do not remove the useless parameters of the normal mode functions, I will use fewer cycles than I remove it.

Reply:
> But the worst things that can happen (including security issues) is when mixing signed and unsigned together.
our audio stream will always be singed type.

Reply:
Since the intrinsic AE_MIN_32_signed is faster than MIN(), so I will use AE_MIN_32_signed, and the arguments of this instruction need to be signed.